### PR TITLE
Re-entrant GroupedGridReduction

### PR DESCRIPTION
**Pull request by naoyam**
Enable re-entrance with `GroupedGridReduction`. Mostly just copied the logic already implemented for `GridReduction` to `GroupedGridReduction`.

See `FusionGroupedReductionChannelsLastBatchNormLike`. The two grid reductions with vectorized iteration domains are grouped as:

```
  #pragma unroll
  for(nvfuser_index_t i241 = 0; i241 < 2; ++i241) {
    // Allocate global tensor T16
    // Allocate global tensor T17
    // Allocate global tensor T18
    reduction::gridReduceGroup<false, true, false, false, true, false, false>(
      T5[i241],
      T15[i241],
      float(0),
      [](float &a, float b) { a = a + b; },
      &T16[0],
      T9[i241],
      T14[i241],
      float(0),
      [](float &a, float b) { a = a + b; },
      &T17[0],
      &T18[0],
      shared_mem,
      true,
      true,
      i241,
      2,
      T19[0],
      T19[1]);
  }
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu [hunk 1]
```diff
@@ -204,10 +204,6 @@ __device__ void gridReduce(
     const nvfuser_index_t n_entrances) {
   T block_reduction_val = init_val;
 
-  // entrance index only matters for non-persistent re-entrant grid reductions.
-  const nvfuser_index_t entrance_ind_ = PERSISTENT_REDUCTION ? 0 : entrance_ind;
-  const nvfuser_index_t n_entrances_ = PERSISTENT_REDUCTION ? 1 : n_entrances;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu:209]: Just removed unused variables

File: torch/csrc/jit/codegen/cuda/test/test_gpu_validator.h [hunk 2]
```diff
@@ -452,6 +454,17 @@ inline void testValidate(
   }
 }
 
+inline void clearL2Cache() {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_validator.h:457]: Just copied from the benchmark directory as I'm doing ad-hoc perf testing in the C++ test files.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp [hunk 3]
```diff
@@ -1272,6 +1272,186 @@ TEST_F(NVFuserTest, FusionPersistentBNBackwardAllreduce_CUDA) {
...
+  auto tv4 = sum(tv3, {0});
+
+  auto tv5 = add(tv2, tv4);
+  fusion.addOutput(tv5);
+
+  groupReductions({tv2, tv4});
+
+  auto tv0_cache = tv0->cacheAfter();
+
+  const int vec = 2;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:1295]: Out of curiosity what does the heuristics select if it's not grouped? Can the heuristics/scheduler run after grouping?

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:1295]: `scheduler_params` output:

```

===== Reduction Stats ========
total_reduction_numel: 99
total_iteration_numel: 999
vectorize_factor: 1
n_tensor_inputs: 1
max_input_dtype_size: 4
block(16, 16, 1)

===== Reduction Parameters ========

Red On Slow Dim

Iteration Domain: blockIdx.x / threadIdx.x / multiple reductions per block /
Inner Reduction Domain: cross block - threadIdx.y / unroll / factor 4
Launch Parameters: BlockDim.x = 16, BlockDim.y = 16, BlockDim.z = -1, GridDim.x = -1, GridDim.y = -1, GridDim.z = -1, Smem Size = 0

====================================


===== Reduction Stats ========
total_reduction_numel: 99
total_iteration_numel: 999
vectorize_factor: 1
n_tensor_inputs: 1
max_input_dtype_size: 4
block(16, 16, 1)

===== Reduction Parameters ========

Red On Slow Dim

Iteration Domain: blockIdx.x / threadIdx.x / multiple reductions per block /
Inner Reduction Domain: cross block - threadIdx.y / unroll / factor 4
Launch Parameters: BlockDim.x = 16, BlockDim.y = 16, BlockDim.z = -1, GridDim.x = -1, GridDim.y = -1, GridDim.z = -1, Smem Size = 0

====================================
```

Grouping before scheduling currently fails as expressions are changed from `ReducitonOp` to `GroupedReductionOp`. Just making it work would be likely just some mechanical changes.

**Pull request merged**